docs/tileentities/tileentity.md

```diff
@@ -0,0 +1,128 @@
+More advanced examples exist in mods, such as quarries, sorting machines, pipes, and displays.
+
+!!! note
+    `TileEntities` aren't a solution for everything and they can cause lag when used in large amounts.
+    When possible, try to avoid them.
+
+## Creating a `TileEntity`
+In order to create a `TileEntity` you need to extend the `TileEntity` class.
+It is important that your `TileEntity` has a default constructor, so that Minecraft can properly load it.
+After you've created your class, you need to register the `TileEntity`. For this you need to call 
+```JAVA
+    GameRegistry#registerTileEntity(Class<? extends TileEntity> tileEntityClass, ResourceLocation key)
+```
+Where the first parameter is simply your TileEntity class and the second the registry name of your TileEntity.
+
+!!! note
+    The method to register a `TileEntity` is using a `String` instead of a `ResourceLocation` before Forge version `14.23.3.2694`
+    This method creates a `ResourceLocation` from the String. Be sure to use the ResourceLocation format modid:tile_entity to avoid it being called minecraft:tile_entity.
+
+## Attaching a `TileEntity` to a `Block`
+To attach your new `TileEntity` to a `Block` you need to override 2 (two) methods within the Block class.
+```JAVA
+    Block#hasTileEntity(IBlockstate state)
+
+    Block#createTileEntity(World world, IBlockState state)
+```
+Using the parameters you can choose if the block should have a `TileEntity` or not.
+Usually you will return `true` in the first method and a new instance of your `TileEntity` in the second method.
+
+At this point the `TileEntity`can be used for Rendering.
+
+## Storing Data within your `TileEntity`
+In order to save data, override the following two methods
+```JAVA
+    TileEntity#writeToNBT(NBTTagCompound nbt)
+
+    TileEntity#readFromNBT(NBTTagCompound nbt)
+```
+These methods are called whenever the `TileEntity` gets (un-)loaded.
+Use them to read and write to the fields in your tile entity class.
+
+!!! warning
+    Whenever your data changes you need to call `TileEntity#markDirty()`, otherwise your `TileEntity` might be skipped while saving.
+
+!!! note
+    It is important that you call the super methods!
+    The tag names `id`, `x`, `y`, `z`, `ForgeData` and `ForgeCaps`tug are reserved by the super methods.
+
+## Keeping a `TileEntity` through changing `BlockStates`
+There might be situations in which you need to change your `BlockState`, an example for this is the vanilla furnace,
+which changes its state from `lit=false` to `lit=true` when fuel and something smeltable is inside.
+Achieving this is rather simple, by overriding following method
+```JAVA
+    TileEntity#shouldRefresh(World world, BlockPos pos, IBlockState oldState, IBlockState newSate)
+```
+!!! note
+    You should actually check the `BlockStates` and not just return `false` in order to prevent unwanted behavior and bugs.
```

Comment:
Maybe expand this a bit to note that this is called whenever the state changes *period* to any other block(state) in the game.